翻译v3.0升级指南 (#1885)
cocos-creator/creator-docs: en/release-notes/upgrade-guide-v3.0.md

# Cocos Creator 3.0 Upgrade Guide

## Version Introduction

__Cocos Creator 3.0__ integrates all the functions of the original __2D__ and __3D__ products, brings many major updates, and will become the main version of __Cocos Creator__. At the same time, 3.0 also continues __Cocos's__ advantages of light weight and high efficiency in 2D categories, and provides an efficient development experience for 3D heavy games.

- **For Cocos Creator 2.x**

>     In order to ensure the smooth transition of an existing __Cocos Creator 2.4__ project, we will use v2.4 as the LTS (long-term support) version and provide continuous updates for the next **two years**! In **2021**, v2.4 will continue to be updated to provide bug fixes and new mini-game platform support to ensure the successful launch of your project; in **2022**, we will continue to provide developers with the key to v2.4 bug fixes to ensure the smooth operation of online games! Therefore,

Reviewer: ```suggestion
    In order to ensure the smooth transition of an existing __Cocos Creator 2.4__ project, Creator v2.4 will be used as the LTS (long-term support) version and provide continuous updates for the next **two years**! In **2021**, v2.4 will continue to be updated to provide bug fixes and new mini-game platform support to ensure the successful launch of your project; in **2022**, we will continue to provide developers with the key to v2.4 bug fixes to ensure the smooth operation of online games! Therefore:
```

    - **Existing 2.x projects can continue to develop without compulsory upgrade to 3.0**.

    - **For new projects, it is recommended to use version 3.0 for development**. We will continue to optimize the development experience and operating efficiency of 3.0 to support the smooth launch of heavy games of different categories such as 2D and 3D.

- **For Cocos Creator 3D**

    The original Cocos Creator 3D, a branch of Creator, has been iterated for a year for Chinese developers, and has successfully launched heavy projects such as **Star Battle** and **Strongest Demon Fighter**! Cocos Creator 3D will be included in v3.0 after the release of Cocos Creator 3.0, and the existing v1.2 projects can be upgraded directly, so Cocos Creator 3D will not be released as a separate version in the future.

__Cocos Creator 3.0__ uses a new future-oriented engine architecture, which will bring high-performance, data-oriented and load-balanced renderers to the engine, and seamlessly support Vulkan & Metal multi-backend rendering. In the future, it will also support mobile VR/AR and some Host platform. 

> For a detailed introduction to the __Cocos Creator 3.0__, please go to [Official Website Update Instructions](https://cocos.com/creator).

Reviewer: ```suggestion
For a detailed introduction to the __Cocos Creator 3.0__, please visit the [Official Website Update Instructions](https://cocos.com/creator).
```

## How to migrate Cocos Creator 2.x projects

> Although **we do not recommend forcing projects in development, especially those about to go live, to upgrade to v3.0**, we have introduced the v2.x resource import tool in Cocos Creator 3.0. This tool supports the perfect import of resources from old projects, as well as assisted migration of code.

Reviewer: ```suggestion
Although **it is not recommended to force projects in development, especially those about to go live, to upgrade to v3.0**, a v2.x resource import tool has been introduced in Cocos Creator 3.0. This tool supports the perfect import of resources from old projects, as well as assistance in migrating code.
```

The example code above, after the code assisted migration by the import plugin, looks like this:

```typescript
// AudioController.ts

import { _decorator, Component, AudioSource } from 'cc';
const { ccclass, property } = _decorator;

@ccclass('AudioController')
export class AudioController extends Component {
    @property
    public audioSource:AudioSource = 'null';

    play () {
        //this.audioSource.play();
    }

    pause () {
        //this.audioSource.pause();
    }

}

/**
 * Note: The original script has been commented out, due to the large changes in the script, there may be missing in the conversion, you need to manually convert it yourself
 */
// cc.Class({
//     extends: cc.Component,
// 
//     properties: {
//         audioSource: {
//             type: cc.AudioSource,
//             default: null
//         },
//     },
// 
//     play: function () {
//         this.audioSource.play();
//     },
// 
//     pause: function () {
//         this.audioSource.pause();
//     },
// 
// });
```

> **Note**.
>

> > 1. If you are converting from JavaScript to TypeScript, you need to declare **all properties** in TypeScript and set the default values.

Reviewer: ```suggestion
> 1. When converting from JavaScript to TypeScript, it is necessary to declare **all properties** in TypeScript and set the default values.
```

> 2. If the **Inspector** panel data is missing, you need to check if the property type is the same as v2.x. > 3.

Reviewer: ```suggestion
2. If the **Inspector** panel data is missing, check if the property type is the same as v2.x. > 3.
```

> 3. If the JavaScript code uses external types, TypeScript prompts: Fix by importing external source files or declarations.

## Quick start for developers of older versions

### Material Upgrade

> In v3.0, we have continued to improve the design of the material system and the built-in Shader API, so when upgrading from v2.x to v3.x, some of the content cannot be upgraded automatically and needs to be adjusted manually by the developer. For details, please refer to the [Material Upgrade Guide](../material-system/effect-2.x-to-3.0.md) documentation.

Reviewer: ```suggestion
In v3.0, there are continued improvements to the design of the material system and the built-in Shader API. When upgrading from v2.x to v3.x, some of the content cannot be upgraded automatically and needs to be adjusted manually by the developer. For details, please refer to the [Material Upgrade Guide](../material-system/effect-2.x-to-3.0.md) documentation.
```

### Engine API Upgrade

#### For Cocos Creator 3D 1.2 users

> 1. Cocos Creator 3.0 asset loading related APIs are consistent with v2.4, both refactored the `loader`. v1.2 users can refer to [v2.4 Asset Manager Upgrade Guide](https://docs.cocos.com/creator/manual/en/release-notes/asset-manager-upgrade-guide.html) for upgrading.

Reviewer: ```suggestion
1. Cocos Creator 3.0 asset loading-related APIs are consistent with v2.4, both refactored the `loader`. v1.2 users can refer to the [v2.4 Asset Manager Upgrade Guide](https://docs.cocos.com/creator/manual/en/release-notes/asset-manager-upgrade-guide.html) when upgrading.
```

- **Dynamic Loading**:

>     When using `bundle.load` or `resources.load` to dynamically load a `sprite-frame` or `texture` in v3.0, the path needs to be specified to a specific sub-resource:

Reviewer 2: ```suggestion
    When using `bundle.load` or `resources.load` to dynamically load a `sprite-frame` or `texture` in v3.0, the path needs to be specified to a specific sub-asset:
```

> 5. Some resources needed for compilation, such as application icons, application startup scripts, etc., v2.4.3 are stored in the build project, while v3.0 are stored in the `native/engine/name of the currently built platform` directory (e.g.: `native/engine/win32`, `native/engine/android`).

Reviewer 2: ```suggestion
5. Some assets needed for compilation, such as application icons, application startup scripts, etc., v2.4.3 are stored in the build project, while v3.0 are stored in the `native/engine/name of the currently built platform` directory (e.g.: `native/engine/win32`, `native/engine/android`).
```

5. Some resources needed for compilation, such as application icons, application startup scripts, etc., v2.4.3 are stored in the build project, while v3.0 are stored in the `native/engine/name of the currently built platform` directory (e.g.: `native/engine/win32`, `native/engine/android`).

## Upgrade FAQ

### After the upgrade, the project script shows errors when opening in VS Code with operations such as binding component definitions

Cocos Creator 3.x enables Strict Mode for TypeScript, which will review the code more strictly and eliminate any problems that may occur due to negligence during the development process.

> If you don't want to use strict mode, you can check **Enable loose mode** in **Project -> Project Settings -> Scripting** in the top menu bar of Creator. As a reminder, we do not encourage turning off strict mode, as strict null checking can reduce some low-level errors in the code runtime.

Reviewer: ```suggestion
To avoid using strict mode, check **Enable loose mode** in **Project -> Project Settings -> Scripting** in the top menu bar of Creator. As a reminder, turning off strict mode is not encouraged as strict null checking can reduce some low-level errors in the code runtime.
```

> For writing specifications in strict mode, you can refer to the official case [Taxi Game 3D](https://github.com/cocos-creator/tutorial-taxi-game).

Reviewer: ```suggestion
For writing specifications in strict mode, please refer to the [Taxi Game 3D](https://github.com/cocos-creator/tutorial-taxi-game) as a test case.
```

### `Action` actions are all disabled

> Because Cocos Creator 3.x removes the `Action` action system and uses the `Tween` tweening system instead.

Reviewer: ```suggestion
Cocos Creator 3.x removes the `Action` action system and uses the `Tween` tweening system instead.
```

### Modifying `size` and `anchor` of 2D nodes does not work

> You need to get the UITransform component on the node first, and then use the corresponding interface, e.g.

Reviewer: ```suggestion
Obtain the UITransform component on the node first, and then use the corresponding interface, e.g.
```

```typescript
const uiTrans = node.getComponent(UITransform)! ;
uiTrans.anchorX = 0.5;
uiTrans.setContentSize(size);
```

### Modifying the `color` of a 2D node does not work

> You need to get the rendering component on the node (e.g. Sprite component) first, and then use the corresponding interface, e.g.

Reviewer: ```suggestion
Obtain the rendering component on the node (e.g. Sprite component) first, and then use the corresponding interface, e.g.
```

```typescript
const uiColor = node.getComponent(Sprite)! ;
uiColor.color = color(255,255,255);
```

### Modifying `skew` for 2D nodes does not work

As of v3.0, the `skew` interface has been removed.

### Grouping is not available, but there are still grouping settings (Layers) in the Creator's project settings panel

The `group` group management in v2.x has been changed to `Layer` since v3.0, as shown below. In v2.x the group name is obtained from `node.group`, while in v3.x the group value is obtained from `node.layer` as **group value**, and the group value is set in exponential power of 2.

![update-setting](update-setting.png)

The layer value of User Layer 0 is: 2<sup>0</sup> = 1.<br>
The layer value of User Layer 1 is: 2<sup>1</sup> = 2.<br> 
The layer value of User Layer 6 is: 2<sup>6</sup> = 64.

### The sibling node set by `zIndex` is invalid

> As of v3.0 the `zIndex` interface has been removed, if you need to reorder the node tree please use the `priority` method instead.

Reviewer: ```suggestion
As of v3.0, the `zIndex` interface has been removed. To reorder the node tree please use the `priority` method instead.
```

### The script mounted on the node is not available via `getComponent()`.

> Please look up the class name of the corresponding script, not the script name, because in v3.x script components are based on the class name defined in the script, not the script name. There are often problems with scripts not being found due to case. Please refer to [Create Script](...) for more details. /scripting/setup.md).

Reviewer: ```suggestion
Please look up the class name of the corresponding script, not the script name. v3.x script components are based on the class name defined in the script, not the script name. There are often problems with scripts not being found due to letter cases. Please refer to the [Create Script](.../scripting/setup.md) documentation for more details.
```

Reviewer: There was a URL formatting error. I fixed it, but please double-check.

Author: Fixed.

> Please look up the class name of the corresponding script, not the script name, because in v3.x script components are based on the class name defined in the script, not the script name. There are often problems with scripts not being found due to case. Please refer to [Create Script](...) for more details. /scripting/setup.md).

Author: ```suggestion
Please look up the class name of the corresponding script, not the script name. v3.x script components are based on the class name defined in the script, not the script name. There are often problems with scripts not being found due to letter cases. Please refer to the [Create Script](../scripting/setup.md) documentation for more details.
```

### Dynamic loading of images in the `resources` folder is not found

> When an image is set to `sprite-frame`, `texture` or other image type, an asset of the corresponding type will be generated in the **Assets** panel. However, if you load `testAssets/image` directly, the type you get will be `ImageAsset` and you must specify the path to the specific child asset.

Reviewer: ```suggestion
When an image is set to `sprite-frame`, `texture` or other image type, an asset of the corresponding type will be generated in the **Assets** panel. However, if loading an `testAssets/image` directly, the type obtained will be `ImageAsset` and it is necessary to specify the path to the specific child asset.
```

> For example, if an image of type `sprite-frame` is set to `testAssets/image` in the `resources` folder, then to load `SpriteFrame` you would write:

Reviewer: ```suggestion
For example, if an image of type `sprite-frame` is set to `testAssets/image` in the `resources` folder, then to load `SpriteFrame` consider the following code:
```

```typescript
resources.load("testAssets/image/spriteFrame", SpriteFrame, (err, spriteFrame) => {
    this.node.getComponent(Sprite).spriteFrame = spriteFrame;
});
```

> If you are loading an image of type `texture`, just change `spriteFrame` to `texture`.

Reviewer: ```suggestion
When loading an image of type `texture`, change `spriteFrame` to `texture`.
```

### The `audioEngine` interface for the audio system is not working and audio cannot be played

Starting from v3.0, the `audioEngine` interface has been removed and the **AudioSource** component is used to control the playback of audio. Please refer to the [AudioSource Component](../audio-system/audiosource.md) documentation for more details.

### `Button` button is not clickable

To rule out code and rendering level problems, please check if the value of `Z` axis in the `Scale` property of the **Button** node is 0, if so, change it to 1.

### The editor is stuck when making changes to the script after upgrade

Check if the property decorator `property` of the component type defined in the script after upgrade is undefined, if not, it is caused by the imported plugin being too old, please refer to [Plugin Upgrade](https://github.com/cocos-creator/plugin-import-2.x) to update the imported plugin Upgrade. After updating the imported plugins, you need to **re-do the project upgrade**.

### When modifying the `Position` of a node in a script after an upgrade, direct changes through the node (e.g. `node.x`) do not take effect

> As of v3.0, direct access to coordinate positions is not allowed on `node` nodes, you need to access `position` before accessing coordinate values. And `position` is a **read-only property** in v3.x. If you need to change it, please use the `setPosition` method. For example:

Reviewer: ```suggestion
As of v3.0, direct access to coordinate positions is not allowed on `node` nodes, it is necessary to access `position` before accessing coordinate values. And `position` is a **read-only property** in v3.x. To change it, please use the `setPosition` method. For example:
```
